subsys/net/lib/download_client/src/download_client.c

```diff
@@ -745,32 +832,50 @@ int download_client_disconnect(struct download_client *const client)
 {
 	int err;
 
-	if (client == NULL || client->fd < 0) {
+	if (client == NULL) {
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&client->mutex, K_FOREVER);
+
+	if (client->fd < 0) {
+		k_mutex_unlock(&client->mutex);
 		return -EINVAL;
 	}
 
 	err = close(client->fd);
```

Comment:
The restructuring is a nice improvement over the old goto-structure. But I still don't like having the `close()` in the disconnect function. I would prefer to move all socket operations including the `close()` to the download_client thread. Putting the mutex around it will even increase the change of locking a thread.
If we just set a disconnect flag in this function, and do everything else in the download_client thread. We might not even need the mutex in this location at all. It would be good to add an event to indicate to the user that the disconnecting is finished. This to prevent them from having to retry the next `download_client_set_host()` until we no longer return `-EALREADY` (or reuse the ERROR or DONE event?).

I don't have a lot of experience with mutexes apart from them being the cause of deadlocks. I would prefer to keep the mutexes as small as possible and only around the variable you are trying to protect, not around functions calls or every element of the struct. The only elements set from outside the thread that might need protection are the URI, config, filename, offset and the disconnect_flag. Do we even need to protect these with a mutex? The worst thing that can happen is someone doing a second download request and us mixing up the file and offset.

Reply:
It makes sense to move `close()` in the `download_thread` to avoid it happening while the `download_thread` is in a `connect()` call, which could be troublesome.

Reply:
Thanks for the proposal.

I'll start implementing the non-blocking disconnect as well.